File: 1085.py
n, k = map(int, input().split())
a = list(map(int, input().split()))

def check(x):
    current_sum = 0
    counter = 0
    for v in a:
        current_sum += v
        if current_sum >= x:
            counter += 1
            current_sum = 0 if current_sum == x else v
            if current_sum > x:
                return False
    
    if current_sum > 0:
        counter += 1

    return counter <= k

result = [10**18]

def dicho(a, l, r, k, res=result):

    if l + 1 >= r:
        return

    m = (l + r) // 2
    if check(m):
        res[0] = m
        dicho(a, l, m, k)
    else:
        dicho(a, m, r, k)


dicho(a, -1, 10**18, k)
print(result[0])

File: 1085.cpp
#include <bits/stdc++.h>
using namespace std;

long long n, k;
vector<long long> a;

bool check(long long x) {
    long long current_sum = 0;
    long long counter = 0;

    for (long long v : a) {
        current_sum += v;

        if (current_sum >= x) {
            counter++;
            if (current_sum == x) {
                current_sum = 0;
            } else {
                current_sum = v;  // reset to v
            }
            if (current_sum > x) {
                return false;
            }
        }
    }

    if (current_sum > 0)
        counter++;

    return counter <= k;
}

long long dicho() {
    long long left = -1;
    long long right = 1000000000000000LL; // 1e15

    while (left + 1 < right) {
        long long middle = (left + right) / 2;
        if (check(middle)) {
            right = middle;
        } else {
            left = middle;
        }
    }

    return right;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> k;
    a.resize(n);
    for (int i = 0; i < n; i++)
        cin >> a[i];

    cout << dicho() << "\n";
    return 0;
}
